File: backend/app/schemas/overlap.py
"""Response schema for the Phase 8 pairwise Fund Overlap Engine."""
from __future__ import annotations

from datetime import date

from pydantic import BaseModel

from app.schemas.funds import DISCLAIMER


class FundRef(BaseModel):
    id: int
    scheme_name: str


class CommonHolding(BaseModel):
    security_name: str
    weight_a: float
    weight_b: float
    min_weight: float


class SectorOverlapDetail(BaseModel):
    sector: str
    weight_a: float
    weight_b: float
    min_weight: float


class OverlapResponse(BaseModel):
    available: bool
    reason: str | None
    fund_a: FundRef
    fund_b: FundRef
    as_of_date_a: date | None
    as_of_date_b: date | None
    common_securities_count: int
    only_in_a_count: int
    only_in_b_count: int
    weighted_overlap_pct: float | None
    sector_overlap_pct: float | None
    overlap_label: str | None
    return_correlation: float | None
    common_holdings: list[CommonHolding]
    sector_detail: list[SectorOverlapDetail]
    disclaimer: str = DISCLAIMER

File: frontend/src/types/overlap.ts
// Mirrors backend/app/schemas/overlap.py

export interface FundRef {
  id: number;
  scheme_name: string;
}

export interface CommonHolding {
  security_name: string;
  weight_a: number;
  weight_b: number;
  min_weight: number;
}

export interface SectorOverlapDetail {
  sector: string;
  weight_a: number;
  weight_b: number;
  min_weight: number;
}

export interface OverlapResponse {
  available: boolean;
  reason: string | null;
  fund_a: FundRef;
  fund_b: FundRef;
  as_of_date_a: string | null;
  as_of_date_b: string | null;
  common_securities_count: number;
  only_in_a_count: number;
  only_in_b_count: number;
  weighted_overlap_pct: number | null;
  sector_overlap_pct: number | null;
  overlap_label: string | null;
  return_correlation: number | null;
  common_holdings: CommonHolding[];
  sector_detail: SectorOverlapDetail[];
  disclaimer: string;
}
